### ISSUE TITLE
Allow displaying a -

### ISSUE THREAD

**Issue by wucke13** [Kixunil/seven_segment#5]
I think it is as well valid use to display a minus. For this, I would replace all the `u8` by `i8` and encode the `-` symbol as `-1`. That way the existing API works as before, but this library can serve as a basis for a multiple seven segment display.

**Comment by Kixunil** [comment]
Good idea, but I'd refrain from overloading `i8` for that, as it'd become confusing. I'd suggest using a newtype called `Symbol`. It can be a struct containing `u8` or an enum (because I'm not sure which is better right now - enum is nicer, but u8 is more optimal, which may be important for embedded).

The `Symbol` type should have these methods:

`from_digit(digit: u8) -> Result<Self, InvalidDigitError>` - checks if it's in range `0-15`
`blank() -> Self` - creates a value representing blank display, I'd suggest using `255` internally
`minus() -> Self` - creates a minus sign, suggested internal value `254`
`is_blank(&self) -> bool` - obvious
`is_minus(&self) -> bool` - obvious
`digit(&self) -> Option<u8>` - returns `Some(x)` if the symbol is digit.

The type should derive `Copy, Clone, Debug, Eq, PartialEq, Hash` and impl Display where digit is displayed directly as number/letter, minus as `-` and blank as ` ` (space).

Additionally, `impl Display for &[Symbol]` could be useful.
Then we could `impl From<u8> for [Symbol; 3]`, `impl From<i8> for [Symbol; 4]`, `impl From<u16> for [Symbol; 5]`, `impl From<i16> for [Symbol; 6]` etc, up to `i128`.

Not sure about `f32` and `f64`, as we don't have point now.

**Comment by Kixunil** [comment]
Possible extension: `fn from_digit_typenum<N: typenum::IsLess<typenum::P16, Output=typenum::B1>>() -> Self` - statically verified literal instatntiation :) conditioned on `typenum` feature (`typenum` as optional dependency)

**Comment by wucke13** [comment]
> It can be a struct containing u8 or an enum

I prefer an `i8`. However, if an enum shall be used, then maybe this:

```
enum Symbol{
    Digit(u8),
    Blank,
    Minus
}
```
It consumes 2 bytes, versus the 1 byte `struct Symbol(u8)`. A byte ain't much, but a 100% size increase just for sweet semantics doesn't sound good too.

> `blank() -> Self` - creates a value representing blank display, I'd suggest using 255 internally

I oppose. IIRC u8 = 255 is bitwise identical to i8 = -1. I suppose using the highest positive value an i8 can inherit (127) as blank and using -1/255 as - sign, it is semantically a better match.
 
> Not sure about f32 and f64, as we don't have point now.

I would ignore this for the time being. Finding out where to put the dot is as simple as dividing through the base. In embedded, people anyways are more likely to use fixed point arithmetic. Maybe in the far future another module can be made with `MultiSevenSegment`. I already wrote a function for the heart of that, determining which digits goes where: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=c8b168ba64fd102ebbaad7136e744cc7

**Comment by Kixunil** [comment]
Hmm, I have some gut feeling that `i8` might suppress some optimizations, but I wouldn't bet on it. Another potential problem I see with `i8` is that it'll need conversion from `u8` that might get accidentally flagged by lame analysis tools (if `as` is used, or `.try_into().unwrap()`). I don't like using signed integers unless they truly, semantically mean "a *number* that can be negative".

It's a shame Rust doesn't have advanced layout optimizations for ranged types yet (hopefully when const generics are stabilized). That's also the reason why I'd start with a struct that has a private field. It could be changed into publicly-visible enum at any moment when it's appropriate without introducing backwards compatibility (although worrying about backwards compatibility at this point might be a bit too much :)).

I don't object to `i8` too much at this point, as it's hidden, so can be changed at any time if there's a problem. In that case, use negative values as special, obviously. `-1` for blank, `-2` for minus sign (or other way around, if you like), could work. Please make those markers consts regardless of what path you choose.

Good point about fixed point arithmetic. Yes, let's ignore `f32` and `f64`.

**Comment by wucke13** [comment]
Well, to me it sounds like you prefer the semantically cleaner way of the enum then? Shall I make a PR for that?

**Comment by Kixunil** [comment]
If this wasn't embedded, I'd prefer that. I have hopes that compiler will be able to optimize it natively one day. So I currently think that just `u8` (or `i8` if it's important for you) encapsulated in a tuple struct is the best for now.

We may backwards-compatibly change it once the optimization lands.

**Comment by Kixunil** [comment]
Please see also #8
